### ISSUE TITLE
Is there any way to apply commands only once by doing a permission check?

### ISSUE THREAD

**Issue by mibby** [elBukkit/MagicPlugin#689]
I have a bunch of wands that I recently migrated to a new unlock GUI. To fix legacy wands to allow people to unlock them without losing their existing, I applied a command that runs permission setting whenever the wand is used. 

Is it possible at all to have a permission check to see if a user has a specific permission before allowing commands to run? That way if they have already been given the permission, it doesn't continue to run the command every time they equip and use the wand.

```
base_hat:
    mode: cast
    active_spell: magichat
    indestructible: true
    locked: true
    keep: true
    bound: true
    effect_color: '888888'
    spells:
        - magichat

hat_example:
    name: Example
    icon: diamond_hoe:38
    inherit: base_hat
    overrides: magichat.command lp user @uuid permission set hat.example
```

**Comment by NathanWolf** [comment]
That's kind of funky- just to throw this out there, should it just consume the wand when they use it and get the permission?  Then they can go grab a migrated version from the GUI?

In any case yes if you want you can use `CheckRequirements`, like

```
conditional:
  icon: stick
  actions:
    cast:
    - class: CheckRequirements
      actions:
      - class: Command
        command: /permission whatever
  parameters:
    target: self
    requirements:
    - permission: hat.example
```

**Comment by mibby** [comment]
I thought about making the old items consumables, but then I'd have hundreds of duplicate wand entries for legacy hats that would eventually phase out over time.

If I'm understanding the ``CheckRequirements`` properly, that would not work because it would only run the command _if_ they meet the permission requirement. I would like the opposite -- interacting with the wand (equipping in this case) runs the permission giving command _unless_ they already have the permission.

**Comment by NathanWolf** [comment]
Oh- whoops, this little tweak *should* work then though

```
conditional:
  icon: stick
  actions:
    cast:
    - class: CheckRequirements
      fail:
      - class: Command
        command: /permission whatever
  parameters:
    target: self
    requirements:
    - permission: hat.example
```

**Comment by NathanWolf** [comment]
Note that `actions` under CheckRequirements changed to `fail`

**Comment by mibby** [comment]
Alright thanks, I'll give it a test later tonight and see how it goes. Though does the wand interact even work when they don't meet the requirement check? I assume the actions only cast when they can interact with the wand and the requirements parameter prevents them from interacting if they don't have that permission.

**Comment by NathanWolf** [comment]
This has nothing to do with wand use, no. The requirements check is part of a spell cast. It's more like a conditional that decides whether or not to run some actions based on some set of conditions. Not really like a "permission check" in the normal sense.

**Comment by mibby** [comment]
@NathanWolf My deepest apologies, I thought I had wrote a reply but I must have not hit sent! Using your conditional check, is there any way to override the requirement permission and the command on the actions on the wand itself?

Previously setup, this worked by providing permissions every time someone interacted with the hat wand, with the wand specifically overriding and providing the correct permission.

spells/magichat.yml
```
magichat:
    icon: diamond_axe{CustomModelData:138}
    icon_disabled: diamond_pickaxe{CustomModelData:138}
    hidden: true
    icon_url: http://textures.minecraft.net/texture/2abee4fe341011b9a0ae2fb83644f695521fd4e4c6198fe23112f1acbf38369
    category: utility
    worth: 100
    actions:
      cast:
      - class: Delay
        delay: 1
      - class: Hat
      - class: Command
    effects:
      cast:
      -  class: EffectSingle
         location: target
         effectlib:
           class: SphereEffect
           iterations: 4
           particle: smoke_large
    parameters:
      target_self: true
      use_item: true
      reason: custom
      cooldown: 1000
      range: 1
      allow_max_range: true
      console: true
      command: lp user @uuid permission set hat.example
```

wands/hats/hat_accessories.yml
```
hat_ice_pendant:
    name: Ice Pendant
    icon: diamond_shovel{CustomModelData:398}
    inherit: base_hat
    overrides: magichat.command lp user @uuid permission set hat.accessories
```

---

Trying your updated method;

spells/magichat.yml
```
magichat:
    icon: diamond_axe{CustomModelData:138}
    icon_disabled: diamond_pickaxe{CustomModelData:138}
    hidden: true
    icon_url: http://textures.minecraft.net/texture/2abee4fe341011b9a0ae2fb83644f695521fd4e4c6198fe23112f1acbf38369
    category: utility
    worth: 100
    actions:
      cast:
      - class: Delay
        delay: 1
      - class: Hat
      - class: CheckRequirements
        fail:
        - class: Command
          command: lp user @uuid permission set hat.example
    effects:
      cast:
      -  class: EffectSingle
         location: target
         effectlib:
           class: SphereEffect
           iterations: 4
           particle: smoke_large
    parameters:
      target_self: true
      use_item: true
      reason: custom
      cooldown: 1000
      range: 1
      allow_max_range: true
      console: true
      requirements:
      - permission: hat.example
```

wands/hats/hat_accessories.yml
```
hat_ice_pendant:
    name: Ice Pendant
    icon: diamond_shovel{CustomModelData:398}
    inherit: base_hat
    overrides: magichat.command lp user @uuid permission set hat.accessories, magichat.permission hat.accessories
```

The updated method doesn't seem to override the command for permission granting or the permission requirement check. Instead it just provides the permission ``hat.example`` from the base spell.

**Comment by NathanWolf** [comment]
Trying using the map-based format of overrides- otherwise I think it's going to lose the spaces, though I'm not sure
```
overrides:
  "magichat.command": "lp user @uuid permission set hat.accessories, magichat.permission hat.accessories"
```

Not really certain if that'll work though and I'm honestly kind of losing track of what all this is trying to do :(

**Comment by mibby** [comment]
Tried both 

``
hat_ice_pendant:
    name: Ice Pendant
    icon: diamond_shovel{CustomModelData:398}
    inherit: base_hat
    overrides:
      "magichat.command": "lp user @uuid permission set hat.accessories"
      "magichat.permission": "hat.accessories"
```

and

```
hat_ice_pendant:
    name: Ice Pendant
    icon: diamond_shovel{CustomModelData:398}
    inherit: base_hat
    overrides:
      "magichat.command": "lp user @uuid permission set hat.accessories, magichat.permission hat.accessories"
```

It still doesn't override and checks for / sets the permission hat.example in the spell config.

> I'm honestly kind of losing track of what all this is trying to do :(
In simple terms, I just want a permission to be given to players on hat equip if they don't already have that permission. Then not constantly re-give the permission if they already have it every time they equip the hat.

For permission checking / setting, I want to set it on a per-hat item basis instead of needing hundreds of different spells attached to hundreds of hats. This would be accomplished by setting overrides on the wand instead of by spell.

**Comment by NathanWolf** [comment]
Ok, so I think I've got a handle on this now.

Problem#1- you need to have "command" in the main parameters section of the magichat spell for it to be overrideable. You can't override things in the "actions" block.

Problem#2- there is no format to be able to override something embedded in a list, like
```
requirements:
 - permission: something
```
Is a list of sections .. there is no way to override that. 

I'm trying to think of a solution to problem #2.

**Comment by NathanWolf** [comment]
Ok I had to make two code changes to get this to work:

1. Add support for dotted paths to sections in wand overrides
2. Add simpler "requirement" parameter to CheckRequirements

And then I had to adjust your spell to move both parameters to the parameters section.  The relevant bits now look like:

spell:
```
magichat2:
    actions:
      cast:
      - class: Delay
        delay: 1
      #- class: Hat
      - class: CheckRequirements
        fail:
        - class: Command
        actions:
        - class: Message
    parameters:
      target_self: true
      message: You have the permission already
      command: lp user @uuid permission set hat.example
      requirement:
        permission: hat.example
```
I added some messaging for testing and cut out all the other stuff but hopefully you get the idea. The wand then looks like:
```
hat_ice_pendant:
    name: Ice Pendant
    icon: diamond_shovel{CustomModelData:398}
    inherit: magichat2
    overrides:
      "magichat2.command": say @p Gave you the permission
      "magichat2.requirement.permission": Magic.commands.mage
```

You'll need the latest dev build, but then it should work!  Sorry for the long wait there.

**Comment by mibby** [comment]
This seems to work wonderfully now, thank you so much!
